src/redprl/refiner.fun

```diff
@@ -1015,8 +1013,190 @@ struct
+          in
+            List.map (fn j => #1 (makeGoal (([(w, P.DIM)], []) |> j)))
+                     (Restriction.Two J ext0 eps0 ext1 eps1)
+          end
+      in
+        listToTel
+          (ListMonad.bind
+             (fn x => ListMonad.bind (intraTube x) group)
+             group)
+      end
+
+    (* Produce the list of goals requiring that tube aspects agree with the cap. *)
+    fun tubeCapGoals H ty r cap group =
+      let
+        fun tubeCap (ext, eps, (u, tube)) =
+          let
+            val J = H >> CJ.EQ ((substSymbol (r,u) tube, cap), ty)
+          in
+            List.map (#1 o makeGoal) (Restriction.One J ext eps)
+          end
+      in
+        listToTel (ListMonad.bind tubeCap group)
+      end
+
+    fun Eq alpha jdg =
+      let
+        val _ = RedPrlLog.trace "HCom.Eq"
+        val H >> CJ.EQ ((lhs, rhs), ty) = jdg
+        val Syn.HCOM (exts0, (r0, r'0), ty0, cap0, tubes0) = Syn.out lhs
+        val () = assertAlphaEq (ty0, ty)
+        val Syn.HCOM (exts1, (r1, r'1), ty1, cap1, tubes1) = Syn.out rhs
+        val () = assertParamEq "HCom.Eq source of direction" (r0, r1)
+        val () = assertParamEq "HCom.Eq target of direction" (r'0, r'1)
+        val _ = ListPair.mapEq (assertParamEq "HCom.Eq extents") (exts0, exts1)
+
+        val (goalTy, _) = makeGoal @@ H >> CJ.EQ_TYPE (ty0, ty1)
+        val (goalCap, _) = makeGoal @@ H >> CJ.EQ ((cap0, cap1), ty)
+
+        val w = alpha 0
+
+        fun interTube ext eps (u, tube0) (v, tube1) =
+          let
+            val tube0 = substSymbol (P.ret w, u) tube0
+            val tube1 = substSymbol (P.ret w, v) tube1
+            val J = H >> CJ.EQ ((tube0,tube1), ty)
+          in
+            List.map (fn j => #1 (makeGoal (([(w, P.DIM)], []) |> j)))
+                     (Restriction.One J ext eps)
+          end
+
+        val group0 = groupTubes exts0 tubes0
+        val group1 = groupTubes exts1 tubes1
+
+        val interTubeGoals =
+          listToTel
+            (ListMonad.bind
+               (fn ((ext, eps, tube0), (_, _, tube1)) => interTube ext eps tube0 tube1)
+               (ListPair.zipEq (group0, group1)))
+      in
+        T.append (T.empty >: goalTy >: goalCap)
+                 (T.append interTubeGoals (T.append (intraTubeGoals H w ty group0)
+                                                    (tubeCapGoals H ty r0 cap0 group0)))
+        #> trivial
+      end
+
+    fun CapEq alpha jdg =
+      let
+        val _ = RedPrlLog.trace "HCom.CapEq"
+        val H >> CJ.EQ ((lhs, rhs), ty) = jdg
+        val Syn.HCOM (exts, (r, r'), ty0, cap, tubes) = Syn.out lhs
+        val () = assertParamEq "HCom.CapEq source and target of direction" (r, r')
+        val () = assertAlphaEq (ty0, ty)
+        val group = groupTubes exts tubes
+
+        val (goalTy, _) = makeGoal @@ H >> CJ.TYPE ty
+        val (goalCap, _) = makeGoal @@ H >> CJ.MEM (cap, ty)
+        val (goalEq, _) = makeGoal @@ H >> CJ.EQ ((cap, rhs), ty)
```

Comment:
So I just added this goal when mechanically transposing the rule, but it looks to me like it makes the previous goal redundant. 

Isn't it the case that if `a = b in A` then `a in A`?